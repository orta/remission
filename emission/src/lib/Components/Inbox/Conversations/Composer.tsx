import * as React from "react"
import { Keyboard, KeyboardAvoidingView, TextInput, TouchableWithoutFeedback } from "react-native"

import styled from "styled-components/native"
import colors from "../../../../data/colors"
import fonts from "../../../../data/fonts"

interface ContainerProps {
  active: boolean
}

const Container = styled.View`
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  border-width: 1;
  border-color: ${colors["gray-regular"]};
  border-radius: 3;
  margin: 0 20px 20px;
  background-color: ${(p: ContainerProps) => (p.active ? "white" : colors["gray-light"])};
`

interface StyledSendButtonProps {
  disabled: boolean
}

const SendButton = styled.Text`
  font-family: ${fonts["avant-garde-regular"]};
  font-size: 12;
  margin-right: 10;
  color: ${(p: StyledSendButtonProps) => (p.disabled ? colors["gray-regular"] : colors["purple-regular"])};
`

interface Props {
  disabled?: boolean
  onSubmit?: (text: string) => any
  value?: string
}

interface State {
  active: boolean
  text: string
}

export default class Composer extends React.Component<Props, State> {
  input?: TextInput | any

  constructor(props) {
    super(props)

    this.state = {
      active: false,
      text: null,
    }
  }

  submitText() {
    Keyboard.dismiss()
    if (this.props.onSubmit) {
      this.props.onSubmit(this.state.text)
      this.setState({ text: null })
    }
  }

  componentDidUpdate() {
    if (this.props.value && !this.state.text) {
      this.setState({ text: this.props.value })
    }
  }

  render() {
    // The TextInput loses its isFocused() callback as a styled component
    const inputStyles = {
      flex: 1,
      fontFamily: fonts["garamond-regular"],
      fontSize: 13,
      paddingLeft: 10,
      paddingTop: 13,
      paddingBottom: 10,
      paddingRight: 10,
    }

    const disableSendButton = !(this.state.text && this.state.text.length) || this.props.disabled

    return (
      <KeyboardAvoidingView behavior="padding" keyboardVerticalOffset={20} style={{ flex: 1 }}>
        {this.props.children}
        <Container active={this.state.active}>
          <TextInput
            placeholder={"Reply..."}
            placeholderTextColor={colors["gray-semibold"]}
            keyboardAppearance={"dark"}
            onEndEditing={() => {
              this.setState({ active: false })
            }}
            onFocus={() => this.setState({ active: this.input.isFocused() })}
            onChangeText={text => this.setState({ text })}
            ref={input => (this.input = input)}
            style={inputStyles}
            multiline={true}
            value={this.state.text}
          />
          <TouchableWithoutFeedback disabled={disableSendButton} onPress={this.submitText.bind(this)}>
            <SendButton disabled={disableSendButton}>SEND</SendButton>
          </TouchableWithoutFeedback>
        </Container>
      </KeyboardAvoidingView>
    )
  }
}
