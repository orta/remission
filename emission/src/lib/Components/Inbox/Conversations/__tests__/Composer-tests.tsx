import * as React from "react"
import "react-native"
import * as renderer from "react-test-renderer"

import Composer from "../Composer"

it("looks correct when rendered", () => {
  const tree = renderer.create(<Composer />).toJSON()
  expect(tree).toMatchSnapshot()
})

describe("regarding the send button", () => {
  it("disables it when there is no text", () => {
    const tree = renderer.create(<Composer />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("disables it even if it contains text if the disabled prop is true", () => {
    const overrideText = "History repeats itself, first as tragedy, second as farce."
    const tree = renderer.create(<Composer value={overrideText} disabled={true} />) as any

    const instance = tree.getInstance()
    instance.componentDidUpdate()

    expect(tree).toMatchSnapshot()
  })
})
