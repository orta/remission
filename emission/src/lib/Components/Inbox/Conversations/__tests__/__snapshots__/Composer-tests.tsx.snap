// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`looks correct when rendered 1`] = `
<View
  keyboardVerticalOffset={20}
  onLayout={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      Object {
        "paddingBottom": 0,
      },
    ]
  }
>
  <View
    active={false}
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#f8f8f8",
          "borderColor": "#e5e5e5",
          "borderRadius": 3,
          "borderWidth": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginBottom": 20,
          "marginLeft": 20,
          "marginRight": 20,
          "marginTop": 0,
        },
        undefined,
      ]
    }
  >
    <TextInput
      keyboardAppearance="dark"
      multiline={true}
      onChangeText={[Function]}
      onEndEditing={[Function]}
      onFocus={[Function]}
      placeholder="Reply..."
      placeholderTextColor="#666666"
      style={
        Object {
          "flex": 1,
          "fontFamily": "AGaramondPro-Regular",
          "fontSize": 13,
          "paddingBottom": 10,
          "paddingLeft": 10,
          "paddingRight": 10,
          "paddingTop": 13,
        }
      }
      value={null}
    />
    <Text
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      allowFontScaling={true}
      disabled={true}
      ellipsizeMode="tail"
      hitSlop={undefined}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "color": "#e5e5e5",
            "fontFamily": "Avant Garde Gothic ITCW01Dm",
            "fontSize": 12,
            "marginRight": 10,
          },
          undefined,
        ]
      }
      testID={undefined}
    >
      SEND
    </Text>
  </View>
</View>
`

exports[`regarding the send button disables it even if it contains text if the disabled prop is true 1`] = `
<View
  keyboardVerticalOffset={20}
  onLayout={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      Object {
        "paddingBottom": 0,
      },
    ]
  }
>
  <View
    active={false}
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#f8f8f8",
          "borderColor": "#e5e5e5",
          "borderRadius": 3,
          "borderWidth": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginBottom": 20,
          "marginLeft": 20,
          "marginRight": 20,
          "marginTop": 0,
        },
        undefined,
      ]
    }
  >
    <TextInput
      keyboardAppearance="dark"
      multiline={true}
      onChangeText={[Function]}
      onEndEditing={[Function]}
      onFocus={[Function]}
      placeholder="Reply..."
      placeholderTextColor="#666666"
      style={
        Object {
          "flex": 1,
          "fontFamily": "AGaramondPro-Regular",
          "fontSize": 13,
          "paddingBottom": 10,
          "paddingLeft": 10,
          "paddingRight": 10,
          "paddingTop": 13,
        }
      }
      value="History repeats itself, first as tragedy, second as farce."
    />
    <Text
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      allowFontScaling={true}
      disabled={true}
      ellipsizeMode="tail"
      hitSlop={undefined}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "color": "#e5e5e5",
            "fontFamily": "Avant Garde Gothic ITCW01Dm",
            "fontSize": 12,
            "marginRight": 10,
          },
          undefined,
        ]
      }
      testID={undefined}
    >
      SEND
    </Text>
  </View>
</View>
`

exports[`regarding the send button disables it when there is no text 1`] = `
<View
  keyboardVerticalOffset={20}
  onLayout={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      Object {
        "paddingBottom": 0,
      },
    ]
  }
>
  <View
    active={false}
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#f8f8f8",
          "borderColor": "#e5e5e5",
          "borderRadius": 3,
          "borderWidth": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginBottom": 20,
          "marginLeft": 20,
          "marginRight": 20,
          "marginTop": 0,
        },
        undefined,
      ]
    }
  >
    <TextInput
      keyboardAppearance="dark"
      multiline={true}
      onChangeText={[Function]}
      onEndEditing={[Function]}
      onFocus={[Function]}
      placeholder="Reply..."
      placeholderTextColor="#666666"
      style={
        Object {
          "flex": 1,
          "fontFamily": "AGaramondPro-Regular",
          "fontSize": 13,
          "paddingBottom": 10,
          "paddingLeft": 10,
          "paddingRight": 10,
          "paddingTop": 13,
        }
      }
      value={null}
    />
    <Text
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      allowFontScaling={true}
      disabled={true}
      ellipsizeMode="tail"
      hitSlop={undefined}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "color": "#e5e5e5",
            "fontFamily": "Avant Garde Gothic ITCW01Dm",
            "fontSize": 12,
            "marginRight": 10,
          },
          undefined,
        ]
      }
      testID={undefined}
    >
      SEND
    </Text>
  </View>
</View>
`
